creusot-contracts/src/ghost_ptr.rs

```diff
@@ -0,0 +1,165 @@
+pub type GhostPtr<T> = *const T;
+
+impl<T: ?Sized> ShallowModel for GhostPtrToken<T> {
+    type ShallowModelTy = FMap<GhostPtr<T>, T>;
+
+    #[logic]
+    fn shallow_model(self) -> Self::ShallowModelTy {
+        *self.0
+    }
+}
+
+impl<T: ?Sized> GhostPtrToken<T> {
+    /// Creates a new [`GhostPtr`] that has no permission
+    #[ensures(result@ == FMap::empty())]
+    pub fn new() -> Self {
+        GhostPtrToken(ghost!(FMap::empty()), PhantomData)
+    }
+
+    #[trusted]
+    #[ensures(forall<ptr1: GhostPtr<T>, ptr2: GhostPtr<T>>
+        self@.contains(ptr1) && self@.contains(ptr2) && ptr1.addr_logic() == ptr2.addr_logic()
+        ==> ptr1 == ptr2)]
+    fn injective_lemma(&self) {}
+
+    #[requires(self@.contains(ptr1) || ptr1 == GhostPtr::null_logic())]
+    #[requires(self@.contains(ptr2) || ptr2 == GhostPtr::null_logic())]
+    #[ensures(result == (ptr1.addr_logic() == ptr2.addr_logic()))]
+    #[ensures(result == (ptr1 == ptr2))]
+    pub fn are_eq(&self, ptr1: GhostPtr<T>, ptr2: GhostPtr<T>) -> bool
+    where
+        T: Sized,
+    {
+        self.injective_lemma();
+        ptr1.addr() == ptr2.addr()
+    }
+
+    /// Leaks memory iff the precondition fails
+    #[requires(self@.is_empty())]
+    pub fn drop(self) {}
+}
+
+impl<T: ?Sized> GhostPtrExt<T> for GhostPtr<T> {
+    #[trusted]
+    #[logic]
+    #[ensures(forall<t: GhostPtrToken<T>> !t@.contains(result))]
+    #[ensures(forall<ptr: GhostPtr<T>> ptr.addr_logic() == result.addr_logic() ==> ptr == result)]
+    fn null_logic() -> Self {
+        absurd
+    }
+
+    #[trusted]
+    #[logic]
+    fn addr_logic(self) -> Int {
+        absurd
+    }
+}
+
+pub trait GhostPtrExt<T: ?Sized>: Sized {
+    #[logic]
+    fn null_logic() -> Self;
+    #[logic]
+    fn addr_logic(self) -> Int;
+}
+
+extern_spec! {
+    impl<T> *const T {
+        // Safety since addr_logic is uninterpreted this just claims ptr::addr is deterministic
+        #[trusted]
+        #[ensures(result@ == self.addr_logic())]
+        fn addr(self) -> usize;
+
+        /// Check if `self` was created with [`GhostPtr::null`]
+        #[trusted]
+        #[ensures(result == (self == GhostPtr::<T>::null_logic()))]
+        fn is_null(self) -> bool;
+    }
+
+    mod std {
+        mod ptr {
+            /// Creates a null [`GhostPtr`] that no GhostToken has permission to
+            // Safety even though this pointer is dangling no GhostToken has permission to it so it's okay
+            #[trusted]
+            #[ensures(result == GhostPtr::<T>::null_logic())]
+            fn null<T>() -> *const T
+            where
+                T: Sized;
+        }
+    }
+}
+
+// Coverts `val` into a [`GhostPtr`] and gives `t` permission to it
+// Safety this pointer was owned by a box so no other GhostToken could have permission to it
+#[trusted]
+#[ensures(!(*t)@.contains(result))]
+#[ensures((^t)@ == (*t)@.insert(result, *val))]
+pub fn ptr_from_box<T>(val: Box<T>, t: &mut GhostPtrToken<T>) -> *const T {
```

Comment:
I kind of feel like we should make these methods on the `GhostPtrToken` type, what do you think? We could also clarify the naming.

Reply:
I made these into methods, let if know if you have good suggestions for the names